addons/purchase/models/purchase_order.py

```diff
@@ -524,6 +524,16 @@ def button_confirm(self):
         return True
 
     def button_cancel(self):
+        purchase_orders_with_receipt = self.filtered(lambda po: any(line.qty_received > 0 for line in po.order_line))
+        purchase_orders_with_invoices = self.filtered(lambda po: any(i.state == 'posted' for i in po.invoice_ids))
+
+        if purchase_orders_with_receipt and purchase_orders_with_invoices:
+            raise UserError(_('Unable to cancel selected order as either some receptions have already been done or Vendor bills created'))
+        if purchase_orders_with_invoices:
+            raise UserError(_('Unable to cancel purchase order %s. You must first cancel the related Vendor bills.', format_list(purchase_orders_with_invoices.mapped('display_name'))))
+        if purchase_orders_with_receipt:
+            raise UserError(_('Unable to cancel purchase order %s as some receptions have already been done.', format_list(purchase_orders_with_receipt.mapped('display_name'))))
```

Comment:
Some English cleanup. Also since having any received products only blocks the orders that have received products, shouldn't this be re-ordered?
```suggestion
        if purchase_orders_with_receipt:
            raise UserError(_("Unable to cancel purchase order(s): %s since some products have already been received. You can only cancel orders with no received product quantities.", format_list(purchase_orders_with_receipt.mapped('display_name'))))
        if purchase_orders_with_invoices:
            raise UserError(_("Unable to cancel purchase order(s): %s. You must first cancel the related vendor bills.", format_list(purchase_orders_with_invoices.mapped('display_name'))))
```

The combined error doesn't make sense imo since not all of the POs will be blocked from being cancelled in that case, but it makes it seem like it is

Reply:
Suggested changes are done.

Thank you.